File: backend/app/models/activity.py
from __future__ import annotations

import uuid
from datetime import date, datetime
from enum import Enum

from sqlalchemy import Date, DateTime, ForeignKey, Integer, String, func
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.models.base import Base


class ResolutionEnum(str, Enum):
    DAY = "day"
    WEEK = "week"
    MONTH = "month"
    YEAR = "year"


class Activity(Base):
    __tablename__ = "activities"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    user_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), ForeignKey("users.id"))
    name: Mapped[str] = mapped_column(String)
    emoji: Mapped[str] = mapped_column(String)
    color: Mapped[str] = mapped_column(String)
    resolution: Mapped[ResolutionEnum] = mapped_column(String)
    archived_at: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now())

    user: Mapped[User] = relationship("User")
    logs: Mapped[list[ActivityLog]] = relationship(back_populates="activity", cascade="all, delete-orphan")


class ActivityLog(Base):
    __tablename__ = "activity_logs"

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    activity_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), ForeignKey("activities.id"))
    timestamp: Mapped[date] = mapped_column(Date)  # log start
    duration_minutes: Mapped[int] = mapped_column(Integer)

    activity: Mapped[Activity] = relationship(back_populates="logs")

File: frontend/src/services/activity.ts
import {baseQuery} from './api';

export const getActivities = () => baseQuery('/activities');

export interface CreateActivityData {
    name: string;
    emoji: string;
    color: string;
    resolution: string;
}

export interface UpdateActivityData {
    name?: string;
    emoji?: string;
    color?: string;
    resolution?: string;
    archived_at?: string | null;
    is_archived?: boolean;
}

export const createActivity = (activity: CreateActivityData) =>
    baseQuery('/activities', {
        method: 'POST',
        body: JSON.stringify(activity),
    });

export const updateActivity = (id: string, updates: UpdateActivityData) =>
    baseQuery(`/activities/${id}`, {
        method: 'PATCH',
        body: JSON.stringify(updates),
    });

export const logTime = (activityId: string, date: string, duration: number) =>
    baseQuery(`/activities/${activityId}/logs`, {
        method: 'POST',
        body: JSON.stringify({timestamp: date, duration_minutes: duration}),
    });
